crypto/evp/evp_utils.c

```diff
@@ -17,31 +17,79 @@
 #include "internal/evp_int.h"    /* evp_locl.h needs it */
 #include "evp_locl.h"
 
+/*
+ * -2 is the returned value from any ctrl function where the control
+ * command isn't supported, and an alternative code path may be chosen.
+ * Since these functions are used to implement ctrl functionality, we
+ * use the same value, and other callers will have to compensate.
+ */
+
 int evp_do_ciph_getparams(const EVP_CIPHER *ciph, OSSL_PARAM params[])
 {
     if (ciph->prov == NULL)
-        return -2;
-    if (ciph->get_params == NULL)
-        return -1;
+        return -2;               /* Unsupported */
+    if (ciph->get_params == NULL) {
+        EVPerr(0, EVP_R_CAN_NOT_GET_PARAMETERS);
+        return 0;
+    }
     return ciph->get_params(params);
 }
 
 int evp_do_ciph_ctx_getparams(const EVP_CIPHER *ciph, void *provctx,
                               OSSL_PARAM params[])
 {
     if (ciph->prov == NULL)
-        return -2;
-    if (ciph->ctx_get_params == NULL)
-        return -1;
+        return -2;               /* Unsupported */
+    if (ciph->ctx_get_params == NULL) {
+        EVPerr(0, EVP_R_CAN_NOT_GET_PARAMETERS);
+        return 0;
+    }
     return ciph->ctx_get_params(provctx, params);
 }
 
 int evp_do_ciph_ctx_setparams(const EVP_CIPHER *ciph, void *provctx,
                               OSSL_PARAM params[])
 {
     if (ciph->prov == NULL)
-        return -2;
-    if (ciph->ctx_set_params == NULL)
-        return -1;
+        return -2;               /* Unsupported */
+    if (ciph->ctx_set_params == NULL) {
+        EVPerr(0, EVP_R_CAN_NOT_SET_PARAMETERS);
+        return 0;
+    }
     return ciph->ctx_set_params(provctx, params);
 }
+
+int evp_do_md_getparams(const EVP_MD *md, OSSL_PARAM params[])
+{
+    if (md->prov == NULL)
+        return -2;               /* Unsupported */
+    if (md->get_params == NULL) {
+        EVPerr(0, EVP_R_CAN_NOT_GET_PARAMETERS);
+        return 0;
+    }
+    return md->get_params(params);
+}
+
+int evp_do_md_ctx_getparams(const EVP_MD *md, void *provctx,
+                            OSSL_PARAM params[])
+{
+    if (md->prov == NULL)
+        return -2;               /* Unsupported */
+    if (md->ctx_get_params == NULL) {
+        EVPerr(0, EVP_R_CAN_NOT_GET_PARAMETERS);
+        return 0;
+    }
+    return md->ctx_get_params(provctx, params);
+}
+
+int evp_do_md_ctx_setparams(const EVP_MD *md, void *provctx,
+                            OSSL_PARAM params[])
+{
+    if (md->prov == NULL)
+        return -2;               /* Unsupported */
+    if (md->ctx_set_params == NULL) {
+        EVPerr(0, EVP_R_CAN_NOT_SET_PARAMETERS);
+        return 0;
+    }
+    return md->ctx_set_params(provctx, params);
+}
```

Comment:
[RANT]

So in this file, we have *six* functions, all almost exactly alike.  Every time I want to make even the slightest change, I must remember to make the change six times.

Is everyone else happy with that?  Really???  I find such duplication and the resulting extra work despicable...  not to mention the obvious proneness to errors whenever someone forgets to make that same change six times...